File: app/src/main/java/com/example/myapplication/data/remote/service/UserAidlRoomService.java
package com.example.myapplication.data.remote.service;

import android.app.Service;
import android.content.Intent;
import android.os.IBinder;

import com.example.myapplication.data.local.room.UserAidlRoomDatabase;
import com.example.myapplication.data.local.room.UserDao;
import com.example.myapplication.data.local.room.UserRoomEntity;
import com.example.myapplication.data.mapper.UserRoomMapper;
import com.example.myapplication.data.remote.aidl.IUserService;
import com.example.myapplication.data.remote.aidl.UserParcelable;

import java.util.ArrayList;
import java.util.List;

public class UserAidlRoomService extends Service {

    private UserDao userDao;

    private final IUserService.Stub binder = new IUserService.Stub() {
        @Override
        public List<UserParcelable> getUsers() {
            List<UserRoomEntity> roomUsers = getUserDao().getUsers();
            List<UserParcelable> users = new ArrayList<>(roomUsers.size());
            for (UserRoomEntity entity : roomUsers) {
                users.add(UserRoomMapper.toParcelable(entity));
            }
            return users;
        }

        @Override
        public boolean addUser(UserParcelable user) {
            if (user == null) {
                return false;
            }
            return getUserDao().insert(UserRoomMapper.toRoomEntity(user)) != -1L;
        }

        @Override
        public boolean updateUser(UserParcelable user) {
            if (user == null) {
                return false;
            }
            return getUserDao().update(UserRoomMapper.toRoomEntity(user)) > 0;
        }

        @Override
        public boolean deleteUser(long userId) {
            return getUserDao().deleteById(userId) > 0;
        }
    };

    @Override
    public IBinder onBind(Intent intent) {
        return binder;
    }

    private UserDao getUserDao() {
        if (userDao == null) {
            userDao = UserAidlRoomDatabase.getInstance(getApplicationContext()).userDao();
        }
        return userDao;
    }
}

File: app/src/main/java/com/example/myapplication/data/remote/service/UserAidlRoomService.kt
package com.example.myapplication.data.remote.service

import android.app.Service
import android.content.Intent
import android.os.IBinder
import com.example.myapplication.data.local.room.UserAidlRoomDatabase
import com.example.myapplication.data.mapper.toParcelable
import com.example.myapplication.data.mapper.toRoomEntity
import com.example.myapplication.data.remote.aidl.IUserService
import com.example.myapplication.data.remote.aidl.UserParcelable

class UserAidlRoomService : Service() {

    private val userDao by lazy(LazyThreadSafetyMode.NONE) {
        UserAidlRoomDatabase.getInstance(applicationContext).userDao()
    }

    private val binder = object : IUserService.Stub() {
        override fun getUsers(): MutableList<UserParcelable> {
            return userDao.getUsers()
                .map { it.toParcelable() }
                .toMutableList()
        }

        override fun addUser(user: UserParcelable?): Boolean {
            user ?: return false
            return userDao.insert(user.toRoomEntity()) != -1L
        }

        override fun updateUser(user: UserParcelable?): Boolean {
            user ?: return false
            return userDao.update(user.toRoomEntity()) > 0
        }

        override fun deleteUser(userId: Long): Boolean {
            return userDao.deleteById(userId) > 0
        }
    }

    override fun onBind(intent: Intent): IBinder = binder
}
